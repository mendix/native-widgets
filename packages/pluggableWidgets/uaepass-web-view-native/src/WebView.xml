<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.webview.UAEPassWebView" supportedPlatform="Native" needsEntityContext="true" offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>UAEPass Web view</name>
    <description>Display an external web page or load custom HTML.</description>
    <studioProCategory>Display</studioProCategory>
    <studioCategory>Display</studioCategory>
    <properties>
        <propertyGroup caption="Data source">
            <property key="url" type="textTemplate" required="false">
                <caption>URL</caption>
                <description>The URL of the web page that is loaded.</description>
            </property>
            <property key="content" type="textTemplate" required="false">
                <caption>Content</caption>
                <description>The HTML content to be rendered. This will precede URL</description>
            </property>
        </propertyGroup>
        <propertyGroup caption="Events">
            <property key="onLoad" type="action" required="false">
                <caption>On load</caption>
                <description />
            </property>
            <property key="onError" type="action" required="false">
                <caption>On error</caption>
                <description />
            </property>
        </propertyGroup>
        <propertyGroup caption="Callback">
            <property key="onMessage" type="action" required="false">
                <caption>On message</caption>
                <description>Used to communicate from the loaded webpage to the native app trough window.ReactNativeWebView.postMessage([string])</description>
            </property>
            <property key="onMessageInput" type="attribute" required="false">
                <caption>Message attribute</caption>
                <description>The attribute where the input message will be stored. This can be used to read the input in the onMessage event.</description>
                <attributeTypes>
                    <attributeType name="String" />
                </attributeTypes>
            </property>
        </propertyGroup>
        <propertyGroup caption="Advanced">
            <property key="userAgent" type="string" required="false">
                <caption>User agent</caption>
                <description>A custom user agent string will be included with the request.</description>
            </property>
            <property key="openLinksExternally" type="boolean" defaultValue="false">
                <caption>Open links externally</caption>
                <description>If true, links embedded in source document will be opened in external browser instead of inside this widget.</description>
            </property>
        </propertyGroup>
    </properties>
</widget>
