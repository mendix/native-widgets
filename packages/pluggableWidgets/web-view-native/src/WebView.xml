<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.webview.WebView" supportedPlatform="Native" needsEntityContext="true" offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Web view</name>
    <description>Display an external web page or load custom HTML.</description>
    <studioProCategory>Display</studioProCategory>
    <studioCategory>Display</studioCategory>
    <properties>
        <propertyGroup caption="Data source">
            <property key="url" type="textTemplate" required="false">
                <caption>URL</caption>
                <description>The URL of the web page that is loaded.</description>
            </property>
            <property key="content" type="textTemplate" required="false">
                <caption>Content</caption>
                <description>The HTML content to be rendered. This will precede URL</description>
            </property>
        </propertyGroup>
        <propertyGroup caption="Events">
            <property key="onLoad" type="action" required="false">
                <caption>On load</caption>
                <description />
            </property>
            <property key="onError" type="action" required="false">
                <caption>On error</caption>
                <description />
            </property>
        </propertyGroup>
        <propertyGroup caption="Callback">
            <property key="onMessage" type="action" required="false">
                <caption>On message</caption>
                <description>Used to communicate from the loaded webpage to the native app trough window.ReactNativeWebView.postMessage([string])</description>
            </property>
            <property key="onMessageInput" type="attribute" required="false">
                <caption>Message attribute</caption>
                <description>The attribute where the input message will be stored. This can be used to read the input in the onMessage event.</description>
                <attributeTypes>
                    <attributeType name="String" />
                </attributeTypes>
            </property>
        </propertyGroup>
        <propertyGroup caption="Advanced">
            <property key="requestMethod" type="enumeration" defaultValue="GET">
                <caption>Request method</caption>
                <description>The method verb for the request</description>
                <enumerationValues>
                    <enumerationValue key="GET">GET</enumerationValue>
                    <enumerationValue key="POST">POST</enumerationValue>
                </enumerationValues>
            </property>
            <property key="postBody" type="textTemplate" required="false">
                <caption>POST Body</caption>
                <description>The body that will be send in the POST request</description>
            </property>
            <property key="headerList" type="object" isList="true" required="false">
                <caption>Header list</caption>
                <description>On Android, headers are only support with the GET method</description>
                <properties>
                    <propertyGroup caption="Header list group">
                        <property key="headerName" type="expression" required="true">
                            <caption>Header name</caption>
                            <description>Name of the header</description>
                            <returnType type="String" />
                        </property>
                        <property key="headerValue" type="expression" required="true">
                            <caption>Header value</caption>
                            <description>Value of the header</description>
                            <returnType type="String" />
                        </property>
                    </propertyGroup>
                </properties>
            </property>
            <property key="userAgent" type="string" required="false">
                <caption>User agent</caption>
                <description>A custom user agent string will be included with the request.</description>
            </property>
            <property key="openLinksExternally" type="boolean" defaultValue="false">
                <caption>Open links externally</caption>
                <description>If true, links embedded in source document will be opened in external browser instead of inside this widget.</description>
            </property>
        </propertyGroup>
    </properties>
</widget>
