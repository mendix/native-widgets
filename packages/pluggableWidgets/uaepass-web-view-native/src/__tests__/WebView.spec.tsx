import { actionValue, dynamicValue } from "@mendix/piw-utils-internal";
import { createElement } from "react";
import { fireEvent, render } from "@testing-library/react-native";
import { WebView as RNWebView } from "react-native-webview";

import { Props, WebView } from "../WebView";

jest.mock("react-native", () => {
    const RN = jest.requireActual("react-native");
    RN.NativeModules.RNCWebView = { isFileUploadSupported: jest.fn(() => true) };
    return RN;
});

jest.mock("react-native-webview", () => {
    const { View } = require("react-native");

    const WebView = (props: any) => {
        return <View {...props} testID="mockWebView" />;
    };

    return { WebView };
});

describe("WebView", () => {
    let defaultProps: Props;

    beforeEach(() => {
        defaultProps = {
            name: "webview-test",
            style: [],
            url: dynamicValue("https://mendix.com"),
            userAgent: "",
            openLinksExternally: false
        };
    });

    it("renders a web view when a url is provided", () => {
        const component = render(<WebView {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders a web view when html content is provided", () => {
        const component = render(<WebView {...defaultProps} content={dynamicValue("Hello, world!")} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders an error when no url or content is provided", () => {
        const component = render(<WebView {...defaultProps} url={dynamicValue()} content={dynamicValue()} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("executes the on load action", () => {
        const onLoadAction = actionValue();
        const component = render(<WebView {...defaultProps} onLoad={onLoadAction} />);

        fireEvent(component.UNSAFE_getByType(RNWebView), "load");

        expect(onLoadAction.execute).toHaveBeenCalledTimes(1);
    });

    it("executes the on error action", () => {
        const onErrorAction = actionValue();
        const component = render(<WebView {...defaultProps} onError={onErrorAction} />);

        fireEvent(component.UNSAFE_getByType(RNWebView), "error");

        expect(onErrorAction.execute).toHaveBeenCalledTimes(1);
    });
});
