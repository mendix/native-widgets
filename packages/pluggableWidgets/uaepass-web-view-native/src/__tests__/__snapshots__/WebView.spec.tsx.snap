// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebView renders a web view when a url is provided 1`] = `
<View
  style={
    {
      "flex": 1,
      "height": "100%",
      "minHeight": 300,
    }
  }
  testID="webview-test"
>
  <View
    onError={[Function]}
    onLoad={[Function]}
    onMessage={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    source={
      {
        "uri": "https://mendix.com",
      }
    }
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
    testID="mockWebView"
    userAgent=""
  />
</View>
`;

exports[`WebView renders a web view when html content is provided 1`] = `
<View
  style={
    {
      "flex": 1,
      "height": "100%",
      "minHeight": 300,
    }
  }
  testID="webview-test"
>
  <View
    onError={[Function]}
    onLoad={[Function]}
    onMessage={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    source={
      {
        "html": "Hello, world!",
      }
    }
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
    testID="mockWebView"
    userAgent=""
  />
</View>
`;

exports[`WebView renders an error when no url or content is provided 1`] = `
<View
  style={{}}
  testID="webview-test"
>
  <Text
    style={
      {
        "color": "red",
        "fontWeight": "bold",
      }
    }
  >
    No URL or content was provided.
  </Text>
</View>
`;
